### PR TITLE
Add support for sending and receiving messages via the ARGO Messaging Service

### PR DESCRIPTION
**Pull request by gregcorbett**
This PR allows the SSM to send messages via the ARGO Messaging Service. It:

- updates the config files to new, optional, settings that, when set, enable the AMS functionality
- allows sending to be authorized with a certificate and key, as opposed to with an AMS token
- provides documentation for setting up a new AMS sender and receiver
- provides documentation for migrating from the EGI message brokers to the AMS
- changes the log format when using AMS, to log the Argo Message ID - **this will require updating the logstash grok pattern used in our ELK monitoring before we deploy**

### CODE REVIEW

File: conf/receiver.cfg [hunk 1]
```diff
@@ -19,6 +26,8 @@ capath: /etc/grid-security/certificates
 check_crls: false
 
 [messaging]
+# Project to which SSM will pull messages from. Uncomment and populate for AMS sending
```
Comment (tofu-rocketry) [inline conf/receiver.cfg]: Move to sep AMS section

File: conf/receiver.cfg [hunk 2]
```diff
@@ -1,3 +1,10 @@
+# SSM protocol/destination type options
+[SSM Type]
+# Either 'MQ-BROKER' for EGI Message Brokers or 'AMS' for Argo Messaging Service
```
Comment (tofu-rocketry) [inline conf/receiver.cfg]: Collapse options to single one for user

File: ssm/ssm2.py [hunk 3]
```diff
@@ -36,6 +36,8 @@
 import time
 import logging
 
+from argo_ams_library import ArgoMessagingService, AmsMessage, AmsException
```
Comment (tofu-rocketry) [inline ssm/ssm2.py]: Prob only want either AMS or STOMP needed to be installed. Use SSL type check.

File: ssm/ssm2.py [hunk 4]
```diff
@@ -55,9 +57,10 @@ class Ssm2(stomp.ConnectionListener):
     REJECT_SCHEMA = {'body': 'string', 'signer':'string?', 'empaid':'string?', 'error':'string'}
     CONNECTION_TIMEOUT = 10
 
-    def __init__(self, hosts_and_ports, qpath, cert, key, dest=None, listen=None,
-                 capath=None, check_crls=False, use_ssl=False, username=None, password=None,
-                 enc_cert=None, verify_enc_cert=True, pidfile=None):
+    def __init__(self, hosts_and_ports, qpath, cert, key, dest=None, listen=None, 
+                 capath=None, check_crls=False, use_ssl=False, username=None, password=None, 
+                 enc_cert=None, verify_enc_cert=True, pidfile=None, protocol="STOMP",
```
Comment (tofu-rocketry) [inline ssm/ssm2.py]: Define as const

File: ssm/ssm2.py [hunk 5]
```diff
@@ -311,6 +333,90 @@ def _send_msg(self, message, msgid):
             # If it fails, use the v3 metod signiture
             self._conn.send(to_send, headers=headers)
 
+    def pull_msg_rest(self):
```
Comment (tofu-rocketry) [inline ssm/ssm2.py]: Pull out AMS and STOMP functions to own files.

File: bin/sender.py [hunk 6]
```diff
@@ -62,40 +62,102 @@ def main():
         sys.exit(1)
     
     log = logging.getLogger('ssmsend')
-    
+
+    # Set defaults for MQ-BROKER only variables
+    use_ssl = None
+    # Set defaults for AMS only variables
+    token = ""
+    project = None
```
Comment (tofu-rocketry) [inline bin/sender.py]: Already ha default in SSM class

File: bin/sender.py [hunk 7]
```diff
@@ -62,40 +62,102 @@ def main():
         sys.exit(1)
     
     log = logging.getLogger('ssmsend')
-    
+
+    # Set defaults for MQ-BROKER only variables
+    use_ssl = None
+    # Set defaults for AMS only variables
+    token = ""
```
Comment (tofu-rocketry) [inline bin/sender.py]: Look at defaulting this in init (currently goes into password)

Comment (tofu-rocketry) [reply bin/sender.py]: Note that empty string is needed as AMS only uses that to define abscence of token.

File: bin/receiver.py [hunk 8]
```diff
@@ -24,6 +24,7 @@
 from ssm import __version__, set_up_logging, LOG_BREAK
 
 from stomp.exception import NotConnectedException
+from argo_ams_library import AmsConnectionException as NotConnectedException
```
Comment (tofu-rocketry) [inline bin/receiver.py]: C.f. just accepting this error rather than aliasing.

File: bin/receiver.py [hunk 9]
```diff
@@ -179,25 +239,31 @@ def main():
...
         # The message listening loop.
         while True:
+            try:
+                time.sleep(1)
+                if protocol == 'HTTPS':
+                    # We need to pull down messages as part of
+                    # this loop when using HTTPS.
+                    ssm.pull_msg_rest()
+
+                if i % REFRESH_DNS == 0:
```
Comment (tofu-rocketry) [inline bin/receiver.py]: Poss doesn't need to be in try...except. Check if affects ELK.

File: ssm/ssm2.py [hunk 10]
```diff
@@ -311,6 +333,90 @@ def _send_msg(self, message, msgid):
...
+                # sent via the SSM, we need to pull down that message
+                # to prevent it blocking the message queue.
+                log.debug("Message %s has no empaid.", msgid)
+                empaid = "N/A"
+            # get the message body
+            body = msg.get_data()
+
+            log.info('Received message. ID = %s, Argo ID = %s', empaid, msgid)
+
+            extracted_msg, signer, err_msg = self._handle_msg(body)
```
Comment (tofu-rocketry) [inline ssm/ssm2.py]: This and bit below duplicated (to 405) from STOMP version. Should factor out to method and just use that. Take care of ack parts for AMS.

File: ssm/ssm2.py [hunk 11]
```diff
@@ -342,14 +450,50 @@ def send_all(self):
...
+                while self._last_msg is None:
+                    if not self.connected:
+                        raise Ssm2Exception('Lost connection.')
+
+                log_string = "Sent %s" % msgid
+
+            elif self._protocol == "HTTPS" and self._dest_type == "AMS":
+                # Then we are sending to an Argo Messaging Service instance.
+                if text is not None:
+                    # First we sign the message
```
Comment (tofu-rocketry) [inline ssm/ssm2.py]: Factor out to an AMS send message function.

File: conf/receiver.cfg (hunk 1 above)
Comment (tofu-rocketry) [reply conf/receiver.cfg]: Actually, I'm just going to change the name of the option and tweak the message.

File: conf/sender.cfg [hunk 12]
```diff
@@ -1,3 +1,11 @@
+################################################################################
```
Comment (tofu-rocketry) [inline conf/sender.cfg]: Apply same changes from receiver.

File: ssm/ssm2.py (hunk 3 above)
Comment (tofu-rocketry) [reply ssm/ssm2.py]: For simplicity of the code, I think we should postpone making it an either/or choice and just require both for the moment.

File: bin/receiver.py (hunk 9 above)
Comment (tofu-rocketry) [reply bin/receiver.py]: Will leave it as there is the `send_ping` bit which could potentially raise an exception.

File: bin/sender.py (hunk 6 above)
Comment (tofu-rocketry) [reply bin/sender.py]: Which is irrelevant as it still needs to be set to pass to the `Ssm2` constructor. However, most of these settings are fetched from config with exception handling if they're not present.

**Comment by tofu-rocketry** [comment]
Pull request opened against GC's branch: gregcorbett/ssm#2

File: ssm/ssm2.py (hunk 5 above)
Comment (tofu-rocketry) [reply ssm/ssm2.py]: Deferred to #95.

File: ssm/ssm2.py (hunk 10 above)
Comment (tofu-rocketry) [reply ssm/ssm2.py]: Deferred to #95.

File: ssm/ssm2.py (hunk 11 above)
Comment (tofu-rocketry) [reply ssm/ssm2.py]: Deferred to #95.

**Comment by tofu-rocketry** [comment]
I've rebased. There were conflicts with the changes from #72 and #93 which I've resolved. Hopefully I haven't made any mistakes. All reviews appreciated.

**Comment by tofu-rocketry** [comment]
@gregcorbett, I can't add you as a reviewer as you opened the PR, but if you could also check my rebasing that would be good.

File: migrating_to_ams.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+The sender configuration is usually found under `/etc/apel/sender.cfg`. Follow the steps below to migrate.
+
+1. Comment out `bdii` and `network`.
+1. Uncomment `host` and set it to `msg-devel.argo.grnet.gr`.
+1. Add the following as a new section at the top of your configuration:
+```
+[sender]
+# Either 'STOMP' for STOMP message brokers or 'AMS' for Argo Messaging Service
+protocol: STOMP
```
Comment (gregcorbett) [inline migrating_to_ams.md]: this should be `protocol: AMS`, as the instruction says:
> Add the following as a new section at the top of your configuration:

Comment (tofu-rocketry) [reply migrating_to_ams.md]: Good spot.

File: bin/receiver.py [hunk 14]
```diff
@@ -103,39 +104,84 @@ def main():
...
-        log.error('Could not connect to LDAP server: %s', e)
-        log.error('System will exit.')
-        log.info(LOG_BREAK)
-        sys.exit(1)    
-    
+
+        # Attempt to configure AMS specific variables.
+        try:
+            token = cp.get('messaging', 'token')
+            project = cp.get('messaging', 'project')
```
Comment (gregcorbett) [inline bin/receiver.py]: This should be `'ams_project'`

File: bin/sender.py [hunk 15]
```diff
@@ -60,42 +60,96 @@ def main():
...
-        log.error('Could not connect to LDAP server: %s', e)
-        log.error('System will exit.')
-        log.info(LOG_BREAK)
-        print 'SSM failed to start.  See log file for details.'
-        sys.exit(1)
-        
+
+        # Attempt to configure AMS project variable.
+        try:
+            project = cp.get('messaging', 'project')
```
Comment (gregcorbett) [inline bin/sender.py]: This should be `'ams_project'`

File: migrating_to_ams.md (hunk 13 above)
Comment (tofu-rocketry) [reply migrating_to_ams.md]: Sorted.

File: bin/receiver.py (hunk 14 above)
Comment (tofu-rocketry) [reply bin/receiver.py]: Corrected.

File: bin/sender.py (hunk 15 above)
Comment (tofu-rocketry) [reply bin/sender.py]: Corrected.

**Comment by tofu-rocketry** [comment]
Any final comments before I merge this?

I'll add the argo-ams-library to the SPEC an setup.py when I'm on the release branch.

**Comment by gregcorbett** [comment]
I am trying to do a test send/receive, currently hitting an error around the `use_ssl` variable.

File: bin/sender.py [hunk 16]
```diff
@@ -130,16 +184,19 @@ def main():
...
-                   capath=cp.get('certificates', 'capath'),
-                   enc_cert=server_cert,
-                   verify_enc_cert=verify_server_cert)
-        
+                      cp.get('messaging', 'path'),
+                      path_type=path_type,
+                      cert=cp.get('certificates', 'certificate'),
+                      key=cp.get('certificates', 'key'),
+                      dest=cp.get('messaging', 'destination'),
+                      use_ssl=use_ssl,
```
Comment (gregcorbett) [inline bin/sender.py]: if using `Ssm2.AMS_MESSAGING`, `use_ssl` doesn't get set so this using it here causes an error.

Comment (tofu-rocketry) [reply bin/sender.py]: My bad. I've re-added setting it to `None`, but only when using AMS this time, for both sender and receiver.

**Comment by gregcorbett** [comment]
I can get both the `sender` and `receiver` to the point where the broker gives me a `401`.

**Pull request merged**